[Obsolete by https://github.com/theforeman/foreman/pull/967] Improve the REMOTE_USER (for example mod_auth_kerb) support

adelton opened the pull request:
This series of patches makes it possible to set up a protected "logon" location in Apache via

  &lt;Location /users/extlogin>
  ...
  &lt;/Location>

and configure mod_auth_\* authentication modules for that location. It was tested with mod_auth_kerb.

The possibilities and issues with the existing REMOTE_USER support are listed at http://projects.theforeman.org/projects/foreman/wiki/Foreman_and_mod_auth_kerb, together with explanations introduced by this pull request.

adelton committed:
Add support for mod_auth_* authentication modules at /users/extlogin.

adelton committed:
If session authenticated via SSO::Apache expired, just visit /users/extlogin again.

adelton committed:
Stop infinite redirection loop to the login_url if we are already there.

This can happen when authorize_login_delegation is true but
/users/extlogin is not configured in Apache to authenticate.

adelton committed:
Fall back to the /users/login method.

Addressing
undefined method `allowed_to?' for nil:NilClass

This can happen when authorize_login_delegation is true but
/users/extlogin is not configured in Apache to authenticate.

adelton committed:
Upon logout, land the user on simple page and offer /users/extlogin logon.

skottler commented:
[test]

ares commented on app/controllers/concerns/foreman/controller/authentication.rb:
```diff
@@ -26,6 +26,7 @@ def authenticate
           session[:original_uri] = request.fullpath
           @available_sso ||= SSO::Base.new(self)
 
+          return if @available_sso.login_url == request.fullpath
```
I think `skip_before_filter :require_login` before filter for UsersController#extlogin already does this. `require_login` just executes authenticate which you skip here. Or am I missing something?

adelton replied:
It seems needed or I get

```
Operation FAILED: Render and/or redirect were called multiple times in this action. Please note that you may only call render OR redirect, and at most once per action. Also note that neither redirect nor render terminate execution of the action, so if you want to exit an action after redirecting, you need to do something like "redirect_to(...) and return".
```

domcleal replied:
@ares the optional_login method calls authenticate, so this gets run.. but then we also render the extlogin page, so need to skip the redirect to the login page below.  So if we're just skipping the one line below if we're already on the login_url page.

adelton replied:
Anyway, my gut feeling is that safeguard like this should be in front of any redirect, unless there is a very specific reason to redirect to the same page.

ares commented on app/controllers/concerns/foreman/controller/authentication.rb:
```diff
@@ -48,6 +49,10 @@ def require_login
+  def optional_login
+    authenticate or redirect_to login_users_path
+  end
+
```
Why do we need to create an extra method for this? Could we just add some "optional_login" or "enable_fallback" flag to SSO method and decide based on that in #authenticate whether to redirect to login_users_path?

adelton replied:
That would likely work too.

ares commented on app/controllers/users_controller.rb:
```diff
@@ -97,6 +98,14 @@ def login
+       login_user(user)
+    end
+  end
+
```
I think this is something, that #authenticate should already do for you.

adelton replied:
It's a simplified copy of login. If it's needed in login, I guess authenticate does not do it. At least that redirect to original_uri is not done by authenticate.

ares commented on app/views/users/extlogout.html.erb:
```diff
@@ -0,0 +1,8 @@
+    <%= image_tag("foreman_white.png", :class=>"login-logo") %>
+  </div>
+  <div class="section">
+    <a href="/users/extlogin">Please log in again.</a>
```
Could you use standard `link_to` helper here with `extlogin_users_path`

adelton replied:
Sure,

```
<%= link_to(_("Please log in again."), extlogin_users_path) %>
```

will work too.

adelton replied:
I've updated this pull request and the change is in d44a9bdbdea928ba63db5c337baef8e57d14f119.

domcleal commented on app/services/sso/apache.rb:
```diff
@@ -18,8 +22,16 @@ def authenticated?
+    end
+
     def logout_url
-      "#{Setting['login_delegation_logout_url']}"
```
I think we need to run this by @eLobato, as he added it only recently.

As per your e-mail, you can't trivially do `Setting['login_delegation_logout_url'] || controller.extlogout_users_path` as the former has a defined default, but perhaps we can change the default to nil in `app/models/setting/auth.rb` so it only returns a true value if a user's set it (our settings differentiate between defaults and set values).

adelton replied:
I tried this but the problem is that the value lives not only in `app/models/setting/auth.rb` but in the `settings` table as well. So on my already installed Foreman, the change in the code does not make any difference. I assume some schema upgrade would be needed and that's beyond my current rails ability.

adelton replied:
When I try to define new settings parameter with nil, I get

```
Validation failed: Default can't be blank (ActiveRecord::RecordInvalid)
  /opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/validations.rb:56:in `save!'
```

adelton replied:
In commit c09a3c39ccd10555eb2b8d73f5858b50608e7f0c I workaround the issue by using '-' as the default value and treat that value as "not set" as well. Maybe something like that could be used for `login_delegation_logout_url` as well, unless there is a better way to specify nil defaults.

domcleal replied:
Can you try adding login_delegation_logout_url to the BLANK_ATTRS constant in app/models/setting.rb?  I think that might permit it to be blank by default.

adelton replied:
I confirm that it works.
Now what is the best way to update the `settings` table value as well?

domcleal replied:
If you change the last argument in app/models/setting/auth.conf then the default value will update in the table.  Only users who have saved the setting at some point will have a "value" stored in the table (which will continue to take precedence).  Most users won't have changed the value, so the change to the default will take effect.

adelton committed:
Use link_to.

domcleal commented:
A couple more general points:
1. Could you [file a new feature](http://projects.theforeman.org/projects/foreman/issues/new) in Redmine to record the motivation behind the change?  Essentially just your PR description.  You can reference this in your commit message as "fixes #9999 - commit message".
2. Some tests are failing, as they check the new routes have appropriate permissions for non-admin users.  For login/logout pages with skipped filters this is inappropriate, so please add the extlogin/logout routes to MAY_SKIP_REQUIRE_LOGIN in test/lib/foreman/access_permissions_test.rb.

ares commented:
Please look at my PR to your branch, I think you reimplement some parts which are already used for other SSO methods. Does it make sense? If you merge it, this PR will be updated as well. https://github.com/adelton/foreman/pull/1

adelton commented:
I've filed http://projects.theforeman.org/issues/3312

adelton committed:
Do not check permissions on extlogin/logout.

Addressing https://github.com/theforeman/foreman/pull/958#issuecomment-26486270 point 2.

adelton commented:
I've updated this pull request and the change for the MAY_SKIP_REQUIRE_LOGIN is in bcaa201c9ff7dad97c521792af1913af761460c9.

adelton commented:
> Please look at my PR to your branch, I think you reimplement some parts which are already used for other SSO methods. Does it make sense? If you merge it, this PR will be updated as well. adelton#1

I experience loops when the authenticated user is not in the database, see https://github.com/adelton/foreman/pull/1#issuecomment-26566848

adelton committed:
fixes #3312 - when authorize_login_delegation_auth_source_user_autocreate is set, creates the unknown authenticated user.

adelton commented:
I've realized c09a3c39ccd10555eb2b8d73f5858b50608e7f0c missed the new app/models/auth_sources/auth_source_external.rb class -- repushed now as c3167dce2c17225a203f798ace3b57dd44901f5d.

domcleal commented on app/views/users/extlogout.html.erb:
```diff
@@ -0,0 +1,8 @@
+<div id="login" class="login-container">
```
Did you want to add a title here, as per #3350?

adelton replied:
Please note that http://projects.theforeman.org/issues/3350 talk about existing code. I have no idea how to affect the html/title from inside of the div.

I assume that once that issue is fixed for users/login and users/logout, the same solution can go to users/extlogout as well.

domcleal replied:
Just add `<% title _("Logout") %>` or similar to the top of this file and it should change it.

adelton replied:
It's not that straightforward -- that also adds

```
    <div class="row control-group">
            <h1 class="span8">Logout</h1>
    </div>
```

to the page, displaying the title on the page as well.

domcleal replied:
Ah ok, pass a second blank parameter and it should disable that:

```
<% title _("Logout"), nil %>
```

adelton replied:
On Mon, Oct 21, 2013 at 12:18:30AM -0700, Dominic Cleal wrote:

> Ah ok, pass a second blank parameter and it should disable that:
> 
> ```
> <% title _("Logout"), nil %>
> ```

It does not seem to have any effect.

## 

Jan Pazdziora
Principal Software Engineer, Identity Management Engineering, Red Hat

domcleal commented on app/models/user.rb:
```diff
@@ -135,6 +135,26 @@ def self.try_to_login(login, password)
+      User.current = User.admin
+      auth_source = AuthSource.find_by_name auth_source_name
+      if not auth_source
+        auth_source = AuthSourceExternal.new(:name => auth_source_name)
```
I don't see the need for the AuthSourceExternal name to be configurable, as we're not allowing the user to manage these objects.  I'm wondering instead if it should be fixed via a class method on AuthSourceExternal in the same way as User.admin.

(The autocreate setting can then become a Boolean instead.)

adelton replied:
On Fri, Oct 18, 2013 at 02:34:09AM -0700, Dominic Cleal wrote:

> I don't see the need for the AuthSourceExternal name to be configurable, as we're not allowing the user to manage these objects.  I'm wondering instead if it should be fixed via a class method on AuthSourceExternal in the same way as User.admin.

You can have multiple different AuthSourceExternal sources. In the
future, the name of the source can well come from the Apache
environment (from the mod_auth_\* or mod_lookup_identity modules) and
it may specify the domain or realm from which this user came.

## 

Jan Pazdziora
Principal Software Engineer, Identity Management Engineering, Red Hat

domcleal replied:
That makes sense, but in that case I still don't see why the user needs a setting for it.  Sounds like something we should handle internally - whether that's fixed for now until we support setting it by realm or otherwise.

adelton replied:
On Fri, Oct 18, 2013 at 02:52:41AM -0700, Dominic Cleal wrote:

> That makes sense, but in that case I still don't see why the user needs a setting for it.  Sounds like something we should handle internally - whether that's fixed for now until we support setting it by realm or otherwise.

Well, they need to specify
authorize_login_delegation_auth_source_user_autocreate
anyway. We can state for now that the value of that parameter has to
be External and check that.

Or do you want the setting to be just boolean? I suspect that will be
harder to change into string later.

## 

Jan Pazdziora
Principal Software Engineer, Identity Management Engineering, Red Hat

domcleal commented on app/controllers/users_controller.rb:
```diff
@@ -97,6 +98,14 @@ def login
+
+  def extlogin
+    if session[:user]
+       user = User.find_by_id(session[:user])
```
Nitpick, 3-space indentation should be 2-space

adelton replied:
On Fri, Oct 18, 2013 at 02:42:50AM -0700, Dominic Cleal wrote:

> Nitpick, 3-space indentation should be 2-space

Good catch, fixed in https://github.com/theforeman/foreman/pull/967.

## 

Jan Pazdziora
Principal Software Engineer, Identity Management Engineering, Red Hat

domcleal commented:
I'll wait for @ares' changes to be sorted out and merged in before doing a thorough review, but I think this is taking good shape.

We'll need some tests to cover it in a few areas, listed below.  Hopefully there are tests already that you can copy and work from, else we'll gladly help out.
- SSO::Apache tests in tests/unit/sso/apache_test.rb for the `authenticated?` behaviour change (and probably more here as @ares moves more stuff into SSO::Apache)
- `User.find_or_create_external_user` in tests/unit/user_test.rb
- ideally too, extlogin/extlogout tests in tests/functional/users_controller_test.rb to check the current user's correctly set (or unset?)
- if we add behaviour to AuthSourceExternal, a unit test for it too

adelton commented:
I have now refactored the first part of this branch (support for /users/extlogin and /users/extlogout, not supporting autocreation of users) as https://github.com/theforeman/foreman/pull/967, including hopefully all comments that we've received so far.

adelton commented:
On Fri, Oct 18, 2013 at 02:49:46AM -0700, Dominic Cleal wrote:

> We'll need some tests to cover it in a few areas, listed below.  Hopefully there are tests already that you can copy and work from, else we'll gladly help out.

I think the tests will have to come from someone more fluent with your
test suite and methodology.

## 

Jan Pazdziora
Principal Software Engineer, Identity Management Engineering, Red Hat

ares commented:
I'll take a look

adelton commented:
Closing this pull request as https://github.com/theforeman/foreman/pull/967 obsoletes it, addressing comments this one gathered from people.
